### PR TITLE
[PT FE] Support setting types/shapes for nested structures

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Some PyTorch models may have `dicts`, `lists`, `tuples` as inputs. We unpack them on `normalize` step in PT FE, but this means that those unpacked inputs do not exist before conversion and `convert_model` can't set shapes/types for them. This PR adds ability for `convert_model` to set shapes/types to any arbitrary input obtained by index or name.*

### Tickets:
 - *CVS-142939*

### CODE REVIEW

File: src/frontends/common/src/input_model.cpp [hunk 1]
```diff
@@ -33,6 +33,18 @@ Place::Ptr InputModel::get_place_by_tensor_name(const std::string& tensor_name)
...
 
+Place::Ptr InputModel::get_place_by_input_index(size_t input_idx) const {
+    if (!m_actual) {
+        return {};
+    }
+    /*const auto& inputs = get_inputs();
+    if (input_idx < inputs.size()) {
+        return inputs.at(input_idx);
+    }
+    return {};*/
```
Comment (rkazants) [inline src/frontends/common/src/input_model.cpp:44]:
```suggestion
```

Comment (rkazants) [reply src/frontends/common/src/input_model.cpp:44]: what will it return for other frontends that have no implementation?

File: src/frontends/pytorch/src/input_model.hpp [hunk 2]
```diff
@@ -45,6 +48,7 @@ class InputModel : public ov::frontend::InputModel {
     std::unordered_map<std::string, std::shared_ptr<frontend::Place>> m_name_to_place;
     std::vector<std::shared_ptr<frontend::Place>> m_inputs;
     std::vector<std::shared_ptr<frontend::Place>> m_outputs;
+    std::vector<std::shared_ptr<frontend::Place>> m_fake_places;
```
Comment (rkazants) [inline src/frontends/pytorch/src/input_model.hpp:51]: may be `m_requested_places`?

File: src/frontends/common/src/input_model.cpp (hunk 1 above)
Comment (mvafin) [reply src/frontends/common/src/input_model.cpp:44]: Will need to add that commented implementation in other frontends. There is no way to do it in base class as I don't see a way

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 3]
```diff
@@ -115,131 +81,194 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
+            for user_shape in user_shapes:
+                if user_shape.get('shape') is not None:
+                    input_model.set_partial_shape(
+                        user_shape['node'], user_shape['shape'])
+                if user_shape.get('data_type') is not None:
+                    data_type = user_shape['data_type']
+                    log.debug('Set data type: {}'.format(data_type))
+                    input_model.set_element_type(user_shape['node'], data_type)
+
+        if freeze_placeholder:
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:220]: As I saw it is always empty, so maybe not.

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 4]
```diff
@@ -115,131 +81,197 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
+            log.debug('Using extract subgraph')
+            new_input_places = [x['node'] for x in user_shapes]
+            new_output_places = [x['node'] for x in outputs]
+            add_names_to_tensors(input_model, user_shapes)
+            input_model.extract_subgraph(new_input_places, new_output_places)
+            # invalidation of existing Place objects could have happened in the operation above
+            if user_shapes:
+                placeholder_shapes = create_target_input_shapes(
+                    new_input_places)
+                print(f"placeholder_shapes: {placeholder_shapes}")
```
Comment (mvafin) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:183]:
```suggestion
```

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 5]
```diff
@@ -115,131 +81,197 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
+            log.debug('Using override_all_outputs')
+            add_names_to_tensors(input_model, user_shapes)
+            new_output_places = [x['node'] for x in outputs]
+            input_model.override_all_outputs(new_output_places)
+            # invalidation of existing Place objects could have happened in the operation above
+            if user_shapes:
+                model_inputs = input_model.get_inputs()
+
+        if user_shapes:
+            print(f"user_shapes: {user_shapes}")
```
Comment (mvafin) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:213]:
```suggestion
```

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 6]
```diff
@@ -115,131 +81,195 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
+        for idx, input_info in enumerate(argv.input):
+            if getattr(input_info, "name", None):
+                place = input_model.get_place_by_tensor_name(input_info.name)
+                if not input_info.shape and not input_info.type:
+                    # If we received place by name, we need to use it for FE to verify
+                    # that such name exist, otherwise we silently ignore it.
+                    # Using dynamic shape should be safe, because FE will not overwrite
+                    # the shape that was produced after conversion, but merge it, so
+                    # dynamic shape will not change anything.
+                    input_model.set_partial_shape(place, PartialShape.dynamic())
```
Comment (popovaan) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:94]: Maybe raise error if the place is not found? Or write a warning?

Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:94]: We can't verify if place is not found here

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 7]
```diff
@@ -115,131 +81,195 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
-                    break
-            if node is None:
-                raise Error("Please check correctness of the command-line. "
-                            "Place (operation or tensor) with name {} is not found.".format(name))
-            place = node.get('node')
-
-            if node.get('data_type'):
-                dtype = node['data_type']
-                ov_type = Type(dtype)
+    if getattr(argv, "framework", None) == "pytorch":
```
Comment (popovaan) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:84]: Why is this branch pytorch specific? We need implementation of `get_place_by_input_index()` in other frontends?

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (popovaan) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:221]: Yes, the freezing is not available in OVC. Please, remove this branch.

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 8]
```diff
@@ -115,131 +81,221 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
+        model_inputs = input_model.get_inputs()
+        def equal_places_lists(list1, list2):
+            if len(list1) != len(list2):
+                return False
+            for p1, p2 in zip(list1, list2):
+                if not p1.is_equal(p2):
+                    return False
+            return True
+        # Currently this only work to reorder inputs/outputs
+        to_override_all_inputs = len(iplaces) > 0 and len(iplaces) <= len(model_inputs) and not equal_places_lists(model_inputs, iplaces)
```
Comment (mvafin) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:112]:
```suggestion
        to_override_all_inputs = len(iplaces) == len(model_inputs) and not equal_places_lists(model_inputs, iplaces)
```

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 9]
```diff
@@ -115,131 +81,221 @@ def add_names_to_tensors(model: InputModel, places: List[Place]):
...
+        to_override_all_outputs = False
+        if argv.output:
+            oplaces = []
+            _outputs = fe_output_user_data_repack(input_model, argv.output, moc_front_end.get_name())
+            for out_desc in _outputs:
+                oplaces.append(out_desc["name"])
+            model_outputs = input_model.get_outputs()
+            to_override_all_outputs = len(oplaces) > 0 and len(oplaces) <= len(model_inputs) and not equal_places_lists(model_outputs, oplaces)
```
Comment (mvafin) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:120]:
```suggestion
            to_override_all_outputs = len(oplaces) == len(model_inputs) and not equal_places_lists(model_outputs, oplaces)
```

File: tools/ovc/openvino/tools/ovc/moc_frontend/pytorch_frontend_utils.py [hunk 10]
```diff
@@ -90,76 +91,84 @@ def get_value_from_list_or_dict(container, name, idx):
     return None
 
 
+def flatten_inputs(inputs, names=None):
+    flattened = []
+    if isinstance(inputs, dict):
```
Comment (eaidova) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pytorch_frontend_utils.py:96]: should we somehow take care in case if we have several dicts with the same keys?

**Review by eaidova — APPROVED** [review]

**Review by rkazants — CHANGES_REQUESTED** [review]
freezing is not needed

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rkazants) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:247]: please remove it

File: src/frontends/tensorflow/src/input_model.cpp [hunk 11]
```diff
@@ -824,6 +824,14 @@ ov::frontend::Place::Ptr InputModel::get_place_by_tensor_name(const std::string&
     return _impl->get_place_by_tensor_name(tensorName);
 }
 
+ov::frontend::Place::Ptr InputModel::get_place_by_input_index(size_t input_idx) const {
+    const auto& inputs = get_inputs();
+    if (input_idx < inputs.size()) {
+        return inputs.at(input_idx);
+    }
+    return {};
```
Comment (rkazants) [inline src/frontends/tensorflow/src/input_model.cpp:832]: we do not need it. Let us throw exception that is not implemented

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py]: removed

File: tools/ovc/openvino/tools/ovc/moc_frontend/pytorch_frontend_utils.py (hunk 10 above)
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pytorch_frontend_utils.py:96]: I am still not sure how we can resolve that. Do we really have such issue in real models?

File: src/frontends/tensorflow/src/input_model.cpp (hunk 11 above)
Comment (mvafin) [reply src/frontends/tensorflow/src/input_model.cpp:832]: This implementation mimics the previous behavior, lets have it here so we can use the same code to process places in future

Comment (rkazants) [reply src/frontends/tensorflow/src/input_model.cpp:832]: For me, it is not safe change. We need to run full validation.
I am leaning to revert it for TF FE because I need to sure that it is not used for TF FE as now

File: tools/ovc/openvino/tools/ovc/moc_frontend/pytorch_frontend_utils.py [hunk 12]
```diff
@@ -90,76 +91,85 @@ def get_value_from_list_or_dict(container, name, idx):
...
 def extract_input_info_from_example(args, inputs):
     try:
         from openvino.frontend.pytorch.utils import pt_to_ov_type_map  # pylint: disable=no-name-in-module,import-error
     except Exception as e:
         log.error("PyTorch frontend loading failed")
         raise e
-    example_inputs = args.example_input
-    data_types = args.placeholder_data_types or {}
-    input_shapes = args.placeholder_shapes or {}
+    example_inputs = args.input_model._example_input if args.input_model._example_input else args.example_input
```
Comment (mvafin) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pytorch_frontend_utils.py:125]:
```suggestion
    example_inputs = args.input_model._example_input if args.input_model._example_input is not None else args.example_input
```

File: src/frontends/tensorflow/src/input_model.cpp (hunk 11 above)
Comment (mvafin) [reply src/frontends/tensorflow/src/input_model.cpp]: ok, removed

**Review by popovaan — APPROVED** [review]

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 13]
```diff
@@ -90,156 +92,164 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
...
-                    break
-            if node is None:
-                raise Error("Please check correctness of the command-line. "
-                            "Place (operation or tensor) with name {} is not found.".format(name))
-            place = node.get('node')
-
-            if node.get('data_type'):
-                dtype = node['data_type']
-                ov_type = Type(dtype)
+    if getattr(argv, "framework", None) == "pytorch":
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:95]: please leave pipeline logic for other FWs unchanged and localize changes only for PyTorch like here:
if pytorrch:
   <new logic>
   return
 
<old code>

Comment (rkazants) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:95]: otherwise, it makes sense to me to run full validation (because precommit weakly covers all ovc cases for different fws).
It is really hard to review.

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 14]
```diff
@@ -72,12 +72,14 @@ def moc_pipeline(argv: argparse.Namespace, moc_front_end: FrontEnd):
...
-    user_shapes, outputs, freeze_placeholder = fe_user_data_repack(
-        input_model, argv.placeholder_shapes, argv.placeholder_data_types,
-        argv.output, {}, moc_front_end.get_name())
+    enabled_transforms, disabled_transforms = get_enabled_and_disabled_transforms()
+    if 'ANALYSIS_JSON_PRINT' in enabled_transforms:
+        # NOTE that model analysis is performed before applying user's settings (inputs's shapes etc.)
+        framework_model = moc_front_end.decode(input_model)
+        json_model_analysis_dump(framework_model)
+        # a model is not processed further in json analysis mode
+        sys.exit(0)
```
[the remark this answers is not in the document]
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:82]: No idea. It was moved unchanged

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 15]
```diff
@@ -90,156 +92,164 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
...
-        """
-        for new_input in places:
-            if 'input_name' not in new_input:
-                continue
-            try:
-                model.add_name_for_tensor(new_input['node'], new_input['input_name'])
-            except NotImplementedFailure as e:
-                # some frontends might not implement this method
-                log.warning('Could not add an additional name to a tensor pointed to by \'{}\'. Details: {}'.format(
-                    new_input['input_name'], str(e)))
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:108]: why did you move it. Please leave it here

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 16]
```diff
@@ -90,156 +92,164 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
...
-    log.debug('Inputs are same: {}, outputs are same: {}'.format(
-        inputs_equal, outputs_equal))
-
-    def create_target_input_shapes(new_input_places):
-        if isinstance(new_input_places, list) and len(new_input_places) > 1 \
-                and isinstance(new_input_places[0], tuple):
-            return new_input_places
-        new_input_place_names = [x.get_names()[0] for x in new_input_places]
-        shapes = [shape for shape in argv.placeholder_shapes.values()]
-        return dict(zip(new_input_place_names, shapes))
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:141]: please leave it here

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 17]
```diff
@@ -90,156 +92,164 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
...
-                log.warning('Could not add an additional name to a tensor pointed to by \'{}\'. Details: {}'.format(
-                    new_input['input_name'], str(e)))
-
-    enabled_transforms, disabled_transforms = get_enabled_and_disabled_transforms()
-    if 'ANALYSIS_JSON_PRINT' in enabled_transforms:
-        # NOTE that model analysis is performed before applying user's settings (inputs's shapes etc.)
-        framework_model = moc_front_end.decode(input_model)
-        json_model_analysis_dump(framework_model)
-        # a model is not processed further in json analysis mode
-        sys.exit(0)
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:116]: why it is moved. please leave it here

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 15 above)
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:108]: It is not moved

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 16 above)
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:141]: It is not moved

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 18]
```diff
@@ -90,156 +92,164 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
...
-            input_model.set_tensor_value(place, value)
-
-    def shape_to_array(shape: PartialShape):
-        return [shape.get_dimension(i) for i in range(shape.rank.get_length())]
+                inputs_equal = True
+
+        outputs_equal = True
+        if outputs:
+            # TODO: Remove this line when new 'cut' helper is introduced
+            if getattr(argv, "framework", None) != "pytorch":
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:186]: why do we check it if this is `else` branch corresponding to non-PyTorch case

Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:186]: Agree, not needed

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 17 above)
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:116]: This code is also needed for pytorch, it is not affected by changes in args.input and other variables so it can be moved in the beginning

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 15 above)
Comment (rkazants) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:108]: It is, now it under else branch. Let us leave it here please

**Review by rkazants — COMMENTED** [review]
please add only changes required for PyTorch and put it under separate if. No need to move helper functions to `else`

**Comment by mvafin** [comment]
@rkazants I moved no helper functions

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 19]
```diff
@@ -90,156 +92,205 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
             [item for item in places_original if any(
                 [item.is_equal(item2['node']) for item2 in places_new])]) == len(places_original)
 
-    def add_names_to_tensors(model: InputModel, places: List[Place]):
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:93]: this function is moved. please leave it here

Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:93]: It is not moved. diff shows it wrong. Please use "hide whitespace" changes feature

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py [hunk 20]
```diff
@@ -90,156 +92,205 @@ def check_places_are_same(places_original: List[Place], places_new: List[Place])
...
+                oplaces.append(out_desc["name"])
+            model_outputs = input_model.get_outputs()
+            to_override_all_outputs = check_places_are_same(model_outputs, [{"node": p} for p in oplaces])
+        if to_override_all_inputs and to_override_all_outputs:
+            input_model.extract_subgraph(iplaces, oplaces)
+        elif to_override_all_inputs:
+            input_model.override_all_inputs(iplaces)
+        elif to_override_all_outputs:
+            input_model.override_all_outputs(oplaces)
+    else:
```
Comment (rkazants) [inline tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:145]: let us do not create `else` branch just call ` moc_front_end.convert(input_model)` in if.
This way you localize PyTorch code.

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 19 above)
Comment (rkazants) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:93]: I don't want to use hide whitespaces. This change is included to all changes

Comment (rkazants) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py:93]: I see no value of this movement from one place to another. Moreover, else branch gets longer by this change

File: tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py (hunk 20 above)
Comment (mvafin) [reply tools/ovc/openvino/tools/ovc/moc_frontend/pipeline.py]: done

**Review by rkazants — APPROVED** [review]
